Implement a WinRM runner

nmaludy opened the pull request:
### Description

This is a runner that utilizes WinRM to connect to a remote Windows host for command and script execution. I've chosen to implement it using the `pywinrm` library (https://github.com/diyan/pywinrm) which seems to be the Python standard, used by tools like Ansible and Salt.

The runner is fairly simplistic, only executing commands on a single host at a time (not multiple in parallel like SSH/remote_runner). Due to some limitations in the `pywinrm` library, some code needed to be copied over in order to support "standard" StackStorm features such as: environment variables, cwd, and timeouts. Ideally we will fork `pywinrm` and create a PR for these changes.

It supports both command and script executions. Currently parameters are not passed to the scripts when executing (in the TODO list)

### Examples

Here are some example actions and commands to show it's working so others can play around:


#### /opt/stackstorm/packs/default/winrm_cmd.yaml
```yaml
---
description: "Run an arbitrary WinRM command"
enabled: true
name: winrm_cmd
pack: default
parameters: {}
runner_type: "winrm-ps-cmd"
```

#### /opt/stackstorm/packs/default/winrm_script.yaml
```yaml
---
description: "Run a WinRM script"
enabled: true
name: winrm_script
entry_point: ps_test.ps1
pack: default
parameters: {}
runner_type: "winrm-ps-script"
```

#### Testing winrm-ps-cmd

Simple test 
```shell
$ st2 run default.winrm_cmd host=hostname.domain.tld username=user@domain.tld password=xxx verify_ssl_cert=false cmd='$PSVersionTable'
..
id: 5b1972fd595497f1ade242cb
status: succeeded
parameters: 
  cmd: $PSVersionTable
  host: hostname.domain.tld
  password: '********'
  username: user@domain.tld
  verify_ssl_cert: false
result: 
  failed: false
  return_code: 0
  stderr: ''
  stdout: "
Name                           Value                                           
----                           -----                                           
PSVersion                      4.0                                             
WSManStackVersion              3.0                                             
SerializationVersion           1.1.0.1                                         
CLRVersion                     4.0.30319.42000                                 
BuildVersion                   6.3.9600.18968                                  
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0}                            
PSRemotingProtocolVersion      2.2                                             


"
```

This proves that bot the `env` and `cwd` options work

```shell
$ st2 run default.winrm_cmd host=hostname.domain.tld username=user@domain.tld password=xxx  verify_ssl_cert=false cwd='C:\\Windows' env='{"NICK_TEST": "VALLLUE"}'  cmd='Get-Location; Get-Item Env:NICK_TEST'
..
id: 5b19772c595497f1ade242f8
status: succeeded
parameters: 
  cmd: Get-Location; Get-Item Env:NICK_TEST
  cwd: C:\Windows
  env:
    NICK_TEST: VALLLUE
  host: hostname.domain.tld
  password: '********'
  username: user@domain.tld
  verify_ssl_cert: false
result: 
  failed: false
  return_code: 0
  stderr: ''
  stdout: "
Path                                                                           
----                                                                           
C:\Windows                                                                     

PSPath        : Microsoft.PowerShell.Core\Environment::NICK_TEST
PSDrive       : Env
PSProvider    : Microsoft.PowerShell.Core\Environment
PSIsContainer : False
Key           : NICK_TEST
Value         : VALLLUE
Name          : NICK_TEST



"
```

This executes the following powershell script `/opt/stackstorm/packs/default/actions/ps_test.ps1`

```powershell
$list_of_things = @("a", "b", "c", "d")

foreach ($thing in $list_of_things) {
        Write-Output $thing
}
```

```shell
$ st2 run default.winrm_script host=hostname.domain.tld username=user@domain.tld password=xxx verify_ssl_cert=false
..
id: 5b19800c595497f1ade24311
status: succeeded
parameters: 
  host: hostname@domain.tld
  password: '********'
  username: user@domain.tld
  verify_ssl_cert: false
result: 
  failed: false
  return_code: 0
  stderr: ''
  stdout: "a
b
c
d
"
  succeeded: true
```

### TODO

- [x] Pass parameters to script executions
- [x] Changelog
- [x] Unit tests
- [x] `core` pack actions
- [x] End-to-end tests
- [x] Documentation
- [x] Examples

LindsayHill commented:
Cool. Resolves https://github.com/StackStorm/st2/issues/1636

nmaludy commented:
### Support for parameters and string escaping

Just pushed support for pulling out named and positional parameters from the `action` metadata file and passing them into the PowerShell script.

Also added string quoting/escaping to help in that effort.

#### /opt/stackstorm/pack/default/actions/winrm_script.yaml

```yaml
---
description: "Run an arbitrary WinRM script"
enabled: true
name: winrm_script
entry_point: ps_test.ps1
pack: default
parameters:
  p_bool:
    type: boolean
    default: true
  p_integer:
    type: integer
    default: 3
  p_number:
    type: number
    default: 12.34
  p_str:
    type: string
    default: "test"
  p_array:
    type: array
    default: ["a", "b", "c"]
  p_obj:
    type: object
    default:
      a: a_value
      b: b_value
  pos0:
    type: string
    position: 0
    default: "pos arg 0"
  pos1:
    type: string
    position: 1
    default: "pos arg 1"
runner_type: "winrm-ps-script"
```

#### /opt/stackstorm/pack/default/actions/ps_test.ps1

```powershell
[CmdletBinding()]
Param(
  [bool]$p_bool,
  [int]$p_integer,
  [double]$p_number,
  [string]$p_str,
  [array]$p_array,
  [hashtable]$p_obj,
  [Parameter(Position=0)]
  [string]$p_pos0,
  [Parameter(Position=1)]
  [string]$p_pos1
)


Write-Output "p_bool = $p_bool"
Write-Output "p_integer = $p_integer"
Write-Output "p_number = $p_number"
Write-Output "p_str = $p_str"
Write-Output "p_array = $($p_array | ConvertTo-Json -Compress)"
Write-Output "p_obj = $($p_obj | ConvertTo-Json -Compress)"
Write-Output "p_pos0 = $p_pos0"
Write-Output "p_pos1 = $p_pos1"
```

#### Execution

```shell
$ st2 run default.winrm_script host=hostname.domain.tld username=user@domain.tld password=xxx verify_ssl_cert=false p_str='"test this `$string"'
..
id: 5b1acbcf595497f1ade243e9
status: succeeded
parameters: 
  host: hostname@domain.tld
  p_str: '"test this `$string"'
  password: '********'
  username: user@domain.tld
  verify_ssl_cert: false
result: 
  failed: false
  return_code: 0
  stderr: ''
  stdout: "p_bool = True
p_integer = 3
p_number = 12.34
p_str = "test this `$string"
p_array = ["a","b","c"]
p_obj = {"a":"a_value","b":"b_value"}
p_pos0 = pos arg 0
p_pos1 = pos arg 1
"
  succeeded: true
```

LindsayHill commented on contrib/runners/winrm_runner/winrm_runner/runner.yaml:
```diff
@@ -0,0 +1,133 @@
+      required: true
+      type: string
+    kwarg_op:
+      default: -
```
Looks like this may need to be quoted, otherwise the runner fails to register with `yaml.scanner.ScannerError: sequence entries are not allowed here`

Seems to allow `--` without quoting, but not `-`.

nmaludy commented:
I realized through my testing that the `requirements.txt` in the `runners` directory wasn't actually being used by anything. 

I just pushed a change that converts over all of the runners to using the same `in-requirements.txt` / `requirements.txt` pattern for "components". 

Now, the `Makefile` will treat runners as other `st2` components and generate a `requirements.txt` file from the `in-requirements.txt` file, consulting the `fixed-requirements.txt` file for version pins.

After this change, the `virtualenv` that is built during unit testing contains the requirements specified by the runners.

An additional change is necessary to `st2-packages` so that these requirements are included in the DEB and RPM packages: https://github.com/StackStorm/st2-packages/pull/562

nmaludy commented:
New update:

- Added lots of unit tests, should be almost full coverage now.

- Added a new runner `winrm-cmd` that executes commands via the Command Prompt instead of PowerShell. There are some commands in Windows that need to be executed on the Command Prompt, so i figured it would be a good option to provide.

- Added new actions to the `core` pack: `core.winrm_cmd` that executes remote Command Prompt commands, along with `core.winrm_ps_cmd` that executes remote PowerShell commands. Both of these are using the new runners.

nmaludy commented:
For anyone interested in testing out the new runner. The steps for using it on an existing system are:

```
git clone https://github.com/nmaludy/st2.git
cd st2
git checkout feature/winrm-runner
cp -r contrib/runners/winrm_runner/
st2ctl reload --register-runners
/opt/stackstorm/st2/bin/pip install pywinrm
```

After this  you can create actions that use the new runners `winrm-cmd`, `winrm-ps-cmd` and `winrm-ps-script`. There are some examples further up in this PR.

Kami commented:
Nice work :+1: 

I will perform a more detailed review shortly.

As far as testing goes - we definitely need some kind of end to end tests which spin up Windows VM and test the functionality end to end, the same as we do for other runners (inside st2tests, etc.).

For that we will need to modify workflows in st2cd, st2ci repos and add new tests to st2tests repo.

Kami commented on contrib/runners/winrm_runner/winrm_runner/winrm_base.py:
```diff
@@ -0,0 +1,301 @@
+
+            try:
+                stdout, stderr, return_code, command_done = \
+                    protocol._raw_get_command_output(shell_id, command_id)
```
How does this function behave? Does it block until it receives output / data? If so, for how long? Is that user configurable?

Is it possible that this function returns immediately in some situations? In this case we probably need to add some `eventlet.sleep(0.2)` or similar to prevent blocking / busy waiting.

Kami commented on contrib/runners/winrm_runner/winrm_runner/winrm_base.py:
```diff
@@ -0,0 +1,301 @@
+            rs = e.response
+            rs.timeout = True
+        # end stackstorm custom
+        session.protocol.cleanup_command(shell_id, command_id)
```
Do we need to run those in `finally` to ensure those two functions are always called?

nmaludy replied:
The `pywinrm` library doesn't, but that doesn't mean it isn't a bad idea!

Kami commented on contrib/runners/winrm_runner/winrm_runner/winrm_base.py:
```diff
@@ -0,0 +1,301 @@
+            ps_str = "$true" if param else "$false"
+        elif isinstance(param, list):
+            ps_str = "@("
+            ps_str += ", ".join([self._param_to_ps(p) for p in param])
```
How do we quote the parameters here? Is this function parameter injection safe? Or we need to do additional quoting? Or it happens somewhere else?

nmaludy replied:
Quoting/escaping is done here: https://github.com/nmaludy/st2/blob/feature/winrm-runner/contrib/runners/winrm_runner/winrm_runner/winrm_base.py#L268-L269

We use the `_multireplace()` function to escape the following sequences: https://github.com/nmaludy/st2/blob/feature/winrm-runner/contrib/runners/winrm_runner/winrm_runner/winrm_base.py#L63-L80

In that dict, all keys are the "source" character, and all values are what they're replaced with.

Kami replied:
Thanks for clarifying.

I see `_multireplace` uses regexes. I hope it's safe :)

Kami commented on tox.ini:
```diff
@@ -52,6 +52,7 @@ commands =
     nosetests --with-timer --rednose -sv contrib/runners/mistral_v2/tests/unit/ contrib/runners/mistral_v2/tests/integration/
     nosetests --with-timer --rednose -sv contrib/runners/python_runner/tests/unit/ contrib/runners/python_runner/tests/integration/
     nosetests --with-timer --rednose -sv contrib/runners/windows_runner/tests/unit/
+    nosetests --with-timer --rednose -sv contrib/runners/winrm_runner/tests/unit/
```
Thanks for adding this here to make sure we also run new tests under Python 3.

It looks like pywinrm also works under Python 3 so that's good as well :+1:

nmaludy commented:
Here are the steps i went through to create a Windows AMI with WinRM enabled:

# Creating AWS Windows Instance

Reference = https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/using-features.customenv.html

* Login to AWS Console
* Navigate to EC2 Console
  * https://console.aws.amazon.com/ec2/
* Click Launch Instance
* Select "Microsoft Windows Server 2016 Base" instance
* Select "t2.micro"
* Click "Next: Configure Instance Details"
* Click "Next: Add Storage"
* Click "Next: Add Tags"
* Click "Next: Configure Security Group"
  * Add a new security rule:
    * Type = RDP
    * Protocol = TCP
    * Port Range = 3389
    * Souce = My IP
  * Add a new security rule:
    * Type = WinRM-HTTPS
    * Protocol = TCP
    * Port Range = 5986
    * Souce = My IP
* Click "Review and Launch"
* Click "Launch"
* < wait 4 minutes so the instance is initialized and password is set >

# Connect to Windows instance
Reference = https://docs.aws.amazon.com/AWSEC2/latest/WindowsGuide/connecting_to_windows_instance.html?icmpid=docs_ec2_console

* Navigate to EC2 Console
  * https://console.aws.amazon.com/ec2/
* Select "Running Instances"
* Select your instance in the list
* Click "Connect"
* Click "Decrypt Password"
* RDP into the host

```shell
xfreerdp -u Administrator -p 'xxx' xxx.us-east-1.compute.amazonaws.com
```

# Configure the host for WinRM remoting
* Start -> PowerShell
* The following will download the Ansible WinRM setup script
  * https://raw.githubusercontent.com/ansible/ansible/devel/examples/scripts/ConfigureRemotingForAnsible.ps1
* Commands

```powershell
   # download the script
   (New-Object System.Net.WebClient).DownloadFile('https://raw.githubusercontent.com/ansible/ansible/devel/examples/scripts/ConfigureRemotingForAnsible.ps1', './ConfigureRemotingForAnsible.ps1')
   # execute the script
   ./ConfigureRemotingForAnsible.ps1
```

# Validate that WinRM works
* Setup

```shell 
virtualenv virtualenv
source ./virtualenv/bin/activate
pip install pywinrm
```

* Test script

```python
#!/usr/bin/env python
import winrm

username = "Administrator"
password = 'xxx'
hostname = "xxx.us-east-2.compute.amazonaws.com"
cmd = "$PSVersionTable"
session = winrm.Session('https://{}:5986/wsman'.format(hostname),
                        auth=(username, password),
                        transport='ntlm',
                        server_cert_validation='ignore')

result = session.run_ps(cmd)
print {'stdout': result.std_out,
       'stderr': result.std_err,
       'exit_status': result.status_code}
```

* Expected output: 

```shell
{'exit_status': 0, 'stderr': '', 'stdout': '\r\nName                           Value                                                                                   \r\n----                           -----                                                                                   \r\nPSVersion                      5.1.14393.2248                                                                          \r\nPSEdition                      Desktop                                                                                 \r\nPSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                 \r\nBuildVersion                   10.0.14393.2248                                                                         \r\nCLRVersion                     4.0.30319.42000                                                                         \r\nWSManStackVersion              3.0                                                                                     \r\nPSRemotingProtocolVersion      2.3                                                                                     \r\nSerializationVersion           1.1.0.1                                                                                 \r\n\r\n\r\n'}
```

# SysPrep the EC2 image
Reference = https://docs.aws.amazon.com/AWSEC2/latest/WindowsGuide/ec2config-service.html#ec2config-sysprep

* Start -> Ec2LaunchSettings
* Click "Shutdown with Sysprep"
* Click "Yes"

# Create a new AMI
* Navigate to EC2 Console
  * https://console.aws.amazon.com/ec2/
* Ensure your instance is stopped
* Click on your instance
* Select Actions -> Image -> Create Image
  * Provide name and description
    * Name = Windows Server 2016 with WinRM
    * Description = Windows Server 2016 with WinRM enabled via Ansible setup script
* Click "Create Image"
* Copy the AMI ID

nmaludy commented:
#### Overview

Did some work on the e2e test pipeline in st2tests, st2ci, st2cd. This _should_ allow us to create a Windows VM and run tests against it. Will need someone from the StackStorm side to help with testing in AWS and full test with everything setup on the st2ci/cd system.


### Pre-requirements

1. Setup the Windows AMI (detailed above)
2. Set a new key/value pair for the AMI ID of the Windows AMI created above `st2 key set ami_windows2016 ami-6a561b15`
3. Create a AWS private key for Windows and put it in `/home/stanley/st2_deploy_windows.pem` (specified as a default in the actions)
4. Install the `windows` pack on the box `st2 pack install windows`

### Associated PRs

https://github.com/StackStorm-Exchange/stackstorm-windows/pull/5 - Adds a new action `wait_for_winrm` so the e2e tests can wait for WinRM login to be available.
https://github.com/StackStorm/st2ci/pull/116 - Links in Windows VM creation in the e2e test pipeline
https://github.com/StackStorm/st2cd/pull/337 - Adds several functions such as creating Windows VMs, decrypting Windows passwords, and launching the Windows tests from e2e workflow
https://github.com/StackStorm/st2tests/pull/137 - Action chain for testing WinRM runner and `core.winrm_xxx` actions

Kami commented:
Thanks for great working on those changes :+1: 

I'm looking into it / testing it today :)

Kami commented:
I performed some basic testing locally (using Python 2 and 3) and everything looks good :+1: 

I tagged this PR with v2.9.0 milestone. This means we will need to wait a bit before we can merge it.

I believe @enykeev will start on the v2.8.0 release process soon. As soon as he unfreezes master for new changes I will merge this and work on the CI and all the related changes.

As discussed on Slack - it would  be great if you can also work on the documentation change - how it works, how to use it and how to configure it :)

Kami commented:
Can you please look into the failing tests - I believe they just need to be updated to take into an account a new runner (orchestra).

Kami commented:
I reverted "from st2common import __version__" change because it needs more work (see https://github.com/StackStorm/st2/pull/4211#issuecomment-403474094 for details).

This way I can continue with merging and testing those changes

Kami commented:
Merged - thanks!

I will continue with merging and testing docs and st2tests PRs tomorrow. It took me way too long today to get this branch in sync and all the tests passing (mostly due to slow CI).
